kratos/utilities/normal_calculation_utils.cpp

```diff
@@ -69,99 +327,116 @@ void NormalCalculationUtils::CalculateOnSimplex(
     const std::size_t Dimension
     )
 {
-    // Resetting the normals
-    const array_1d<double,3> zero = ZeroVector(3);
-
-    if (rModelPart.GetCommunicator().GetDataCommunicator().IsDistributed()) {
-        // If Parallel make sure normals are reset in all partitions
-        VariableUtils().SetFlag(VISITED, false, rModelPart.Nodes());
+    // Initialize the normals
+    InitializeNormals<Condition>(rModelPart);
+
+    // Calling CalculateOnSimplex for conditions
+    const auto& r_process_info = rModelPart.GetProcessInfo();
+    const bool has_domain_size = r_process_info.Has(DOMAIN_SIZE);
+    KRATOS_ERROR_IF(has_domain_size && Dimension == 0) << "Dimension not defined" << std::endl;
+    const SizeType dimension_in_model_part = has_domain_size ? r_process_info.GetValue(DOMAIN_SIZE) : Dimension;
+    KRATOS_WARNING_IF("NormalCalculationUtils", dimension_in_model_part != Dimension) << "Inconsistency between DOMAIN_SIZE and Dimension provided" << std::endl;
+    this->CalculateOnSimplex(rModelPart.Conditions(), dimension_in_model_part);
+
+    // Synchronize the normal
+    rModelPart.GetCommunicator().AssembleCurrentData(NORMAL);
+}
 
-        for(auto& r_cond : rModelPart.Conditions()) {
-            for(auto& r_node: r_cond.GetGeometry()) {
-                r_node.Set(VISITED, true);
-            }
-        }
+/***********************************************************************************/
+/***********************************************************************************/
 
-        rModelPart.GetCommunicator().SynchronizeOrNodalFlags(VISITED);
+void NormalCalculationUtils::SwapNormals(ModelPart& rModelPart)
+{
+    KRATOS_TRY
 
-        for(auto& r_node: rModelPart.Nodes()) {
-            if(r_node.Is(VISITED)) {
-                r_node.FastGetSolutionStepValue(NORMAL) = zero;
-            }
-        }
-    } else {
-        // In serial iteratre normally over the condition nodes
-        for(auto& r_cond: rModelPart.Conditions()) {
-            for(auto& r_node: r_cond.GetGeometry()) {
-                r_node.FastGetSolutionStepValue(NORMAL) = zero;
-            }
-        }
+    for(auto& r_cond : rModelPart.Conditions()) {
+        GeometryType& r_geometry = r_cond.GetGeometry();
+        Node<3>::Pointer paux = r_geometry(0);
+        r_geometry(0) = r_geometry(1);
+        r_geometry(1) = paux;
     }
 
-    this->CalculateOnSimplex(rModelPart.Conditions(),Dimension);
-    rModelPart.GetCommunicator().AssembleCurrentData(NORMAL);
+    KRATOS_CATCH("")
 }
 
 /***********************************************************************************/
 /***********************************************************************************/
 
-void NormalCalculationUtils::SwapNormals(ModelPart& rModelPart)
+void NormalCalculationUtils::ComputeUnitNormalsFromAreaNormals(ModelPart& rModelPart)
 {
-    KRATOS_TRY
+    // We iterate over nodes
+    auto& r_nodes_array = rModelPart.Nodes();
```

Comment:
```suggestion
    auto& r_nodes_array = rModelPart.GetCommunicator().LocalMesh().Nodes();
```

otherwise it won't work in MPI